lib/broadway.ex

```diff
@@ -950,7 +966,11 @@ defmodule Broadway do
     metadata = Map.new(Keyword.get(opts, :metadata, []))
 
     ref = make_ref()
-    ack = {Broadway.CallerAcknowledger, {self(), ref}, :ok}
+
+    ack_module = Keyword.get(opts, :acknowledger_module, Broadway.CallerAcknowledger)
+    ack_data = Keyword.get(opts, :acknowledger_data, :ok)
```

Comment:
I would prefer only `:acknowledger` as it was, these details are too granular. :)

Reply:
So I should not bother setting the data?

Reply:
What if the acknowledger is a function that receives the data and returns the acknowledger tuple? 🤔

Reply:
I don't fully understand what is the role of data in the acknowledger. Is it just a piece of state that can be carried over with each message till the messages are acknowledged? If my understanding is correct then the generator approach might be a good one, but that function would take `ack_ref` (not data) and output that tuple.

When I used to work on the Membrane project we used to always focus on users when designing APIs. I think that approach with the function has a small flaw. It requires a programmer that calls this to know how this tuple looks and know that there needs to be `ack_ref`. Providing module and data as separate options has a drawback that someone might for some reason provide data only and get confused about what is happening.

All in all, maybe it should be just one option and either:
  - a function as you mentioned
  - module | {module, data}

It might sound weird but this conversation made me realize that I might have made a bad design decision in the work that originated this (I do still think this feature would be awesome). Thanks for being 🦆 .

Reply:
I think a function that receives data+make_ref will work and be the most general approach we can start with. :)

Reply:
> Is it just a piece of state that can be carried over with each message till the messages are acknowledged?

The idea is that you may want to generate acknowledger info specific to a certain data. I don't have a use case but just trying to be general as possible on the first pass.

Maybe the function is too low-level but I prefer to start with the low-level but general approach and then we can provide higher-level conveniences on top!

Reply:
> What if the acknowledger is a function that receives the data and returns the acknowledger tuple? 🤔

Would make sense to pass the `ref` (and possibly the `pid`) to the acknowledger function as well?

Something like this:
```elixir
Broadway.test_message(MyPipeline, "", acknowledger: fn data, {pid, ref} ->
  {CallerAcknowledger,  {pid, ref}, data}
end)
```

Regardless, we should probably document that the `:ack` message in the example test is sent by `CallerAcknowledger`.

Reply:
Wouldn't the data be always the same? In function I've edited it is hardcoded to be `:ok` tuple.

Reply:
In this case it would be the data of each individual message (i.e. each element in the Enum.map below).

Reply:
That makes sense to me in the case of `test_batch/3` but does not make a single bit of sense to me in the case of `test_message/3`, because in that case user is providing exactly a single piece of data and that function would receive it.

I can see this as a superior solution though. If we agree that the concern voiced above is not an issue, I will implement this solution. Otherwise, I am open to discussion.

Reply:
Yeah, let's start with the function for both for consistency. This will solve the problem, and we figure out how to make it more convenient alter. :)